components/core/src/clp_s/JsonParser.cpp

```diff
@@ -557,6 +621,292 @@ int32_t JsonParser::add_metadata_field(std::string_view const field_name, NodeTy
     return m_archive_writer->add_node(metadata_subtree_id, type, field_name);
 }
 
+auto JsonParser::get_archive_node_type(
+        clp::ffi::SchemaTree::Node::Type ir_node_type,
+        bool node_has_value,
+        std::optional<clp::ffi::Value> const& node_value
+) -> NodeType {
+    // figure out what type the node is in archive node type
+    NodeType archive_node_type = NodeType::Unknown;
+    switch (ir_node_type) {
+        case clp::ffi::SchemaTree::Node::Type::Int:
+            archive_node_type = NodeType::Integer;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Float:
+            archive_node_type = NodeType::Float;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Bool:
+            archive_node_type = NodeType::Boolean;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::UnstructuredArray:
+            archive_node_type = NodeType::UnstructuredArray;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Str:
+            if (node_value && node_value->is<std::string>()) {
+                archive_node_type = NodeType::VarString;
+            } else {
+                archive_node_type = NodeType::ClpString;
+            }
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Obj:
+            if (node_has_value) {
+                if (node_value->is_null()) {
+                    archive_node_type = NodeType::NullValue;
+                } else {
+                    archive_node_type = NodeType::Object;
+                }
+            } else {
+                archive_node_type = NodeType::Object;
+            }
+            break;
+        default:
+            break;
+    }
+    return archive_node_type;
+}
+
+auto JsonParser::get_archive_node_id(
+        std::unordered_map<uint32_t, std::vector<std::pair<NodeType, int32_t>>>&
```

Comment:
1.  The hash function used by the unordered map can't handle hashing a pair. We could write a custom hash function, but I could use some direction with what is the best practice for a custom hash function to avoid conflicts. The vector should be at the worst case 2 long, since in the worst case a ir schema node could translate to two different archive types. Originally I had a map<tuple<ir node id, NodeType>, archive node id>, but I was advice to change to this for better performance. 

2.  Since it's only used for the IR->archive pathway I'm inclined to leave it as is.  As for the name I kind of like having the structure type/ purpose recognized in the name, because when it's used I think it makes the action read a bit more like a sentence. Maybe ir_node_to_archive_node_id_translations or ir_node_to_archive_node_id_mapping?